sbom/spdx/encoder.go

```diff
@@ -0,0 +1,444 @@
+		PackageDownloadLocation: "NOASSERTION",
+	}
+	return pkg
+}
+
+func newSpdxPackageFromDistribution(d *claircore.Distribution) *v2_3.Package {
+	var extRefs []*v2_3.PackageExternalReference
+
+	if d.CPE.String() != "" {
+		extRefs = append(extRefs, &v2_3.PackageExternalReference{
+			Category: "SECURITY",
+			RefType:  "cpe23Type",
+			Locator:  d.CPE.String(),
+		})
+	}
+
+	if d.DID != "" {
+		extRefs = append(extRefs, &v2_3.PackageExternalReference{
+			Category: "OTHER",
+			RefType:  "did",
+			Locator:  d.DID,
+		})
+	}
+
+	if d.VersionID != "" {
+		extRefs = append(extRefs, &v2_3.PackageExternalReference{
+			Category: "OTHER",
+			RefType:  "version_id",
+			Locator:  d.VersionID,
+		})
+	}
+
+	if d.PrettyName != "" {
+		extRefs = append(extRefs, &v2_3.PackageExternalReference{
+			Category: "OTHER",
+			RefType:  "pretty_name",
+			Locator:  d.PrettyName,
+		})
+	}
+
+	dist := &v2_3.Package{
+		PackageName:               d.Name,
+		PackageSPDXIdentifier:     v2common.ElementID("Distribution-" + d.ID),
+		PackageVersion:            d.Version,
+		PackageDownloadLocation:   "NOASSERTION",
+		FilesAnalyzed:             true,
+		PackageExternalReferences: extRefs,
+		PackageSummary:            "distribution",
+		PrimaryPackagePurpose:     "OPERATING-SYSTEM",
+	}
+
+	return dist
+}
+
+func newSpdxPackageFromRepository(r *claircore.Repository) *v2_3.Package {
+	var extRefs []*v2_3.PackageExternalReference
+
+	if r.CPE.String() != "" {
+		extRefs = append(extRefs, &v2_3.PackageExternalReference{
+			Category: "SECURITY",
+			RefType:  "cpe23Type",
+			Locator:  r.CPE.String(),
+		})
+	}
+
+	if r.URI != "" {
+		extRefs = append(extRefs, &v2_3.PackageExternalReference{
+			Category: "OTHER",
+			RefType:  "uri",
+			Locator:  r.URI,
+		})
+	}
+
+	if r.Key != "" {
+		extRefs = append(extRefs, &v2_3.PackageExternalReference{
+			Category: "OTHER",
+			RefType:  "key",
+			Locator:  r.Key,
+		})
+	}
+
+	repo := &v2_3.Package{
+		PackageName:               r.Name,
+		PackageSPDXIdentifier:     v2common.ElementID("Repository-" + r.ID),
+		PackageDownloadLocation:   "NOASSERTION",
+		FilesAnalyzed:             true,
+		PackageSummary:            "repository",
+		PackageExternalReferences: extRefs,
+		PrimaryPackagePurpose:     "OTHER",
+	}
+
+	return repo
+}
+
+func cmpRelationship(a, b *v2_3.Relationship) int {
+	refBCpm := strings.Compare(string(a.RefB.ElementRefID), string(b.RefB.ElementRefID))
+	if refBCpm != 0 {
+		return refBCpm
+	}
+
+	return strings.Compare(a.Relationship, b.Relationship)
+}
+
+// getVersion will attempt to read out the current binary's debug info, find the
+// claircore version (this was copied from Clair).
+func getVersion() string {
+	var core string
+	info, infoOK := debug.ReadBuildInfo()
+	if infoOK {
+		for _, m := range info.Deps {
+			if m.Path != "github.com/quay/claircore" {
+				continue
+			}
+			core = m.Version
+			if m.Replace != nil && m.Replace.Version != m.Version {
+				core = m.Replace.Version
+			}
```

Comment:
are we able to break here, or would we have anything to potentially overwrite this with after the first time? Similarly, can we just return here instead? and then just default return "unknown revision" outside the loop